### PR TITLE
use conf to get num cores per executor

### PR DESCRIPTION
**Pull request by imatiach-msft**
We should use a combination of spark conf and number of cores per machine via java.lang.Runtime.getRuntime.availableProcessors to figure out the number of cores per executor.
Fixed based on discussion here: https://github.com/Azure/mmlspark/issues/379
Thank to @humbinal for figuring out how to fix the issue.
Also should resolve the issue brought by @ywskycn here: https://github.com/Azure/mmlspark/issues/346

Note: This may be eventually resolved in a better way with spark's barrier execution mode when it is implemented aka project hydrogen (without needing to compute num cores):
https://jira.apache.org/jira/browse/SPARK-24374

### CODE REVIEW

**Comment by ywskycn** [comment]
@imatiach-msft I think we also need to include "spark.task.cpus" which is the number of cores to allocate for each task. "spark.executor.cores"/"spark.task.cpus" is the number of tasks in one executor.

**Comment by imatiach-msft** [comment]
@ywskycn done!  I used spark.executer.cores/spark.task.cpus.  It looks like task cpus is usually default to 1 anyway (or not defined, as on my databricks cluster).  Hopefully this will handle all cluster scenarios (other than possibly the spark standalone cluster with different # of cores per executor, which I haven't encountered yet from users).

**Comment by ywskycn** [comment]
LGTM

File: src/lightgbm/src/main/scala/LightGBMUtils.scala [hunk 1]
```diff
@@ -133,7 +133,13 @@ object LightGBMUtils {
   def getNumCoresPerExecutor(dataset: Dataset[_]): Int = {
     val spark = dataset.sparkSession
     import spark.implicits._
-    spark.range(0, 1).map(_ => java.lang.Runtime.getRuntime.availableProcessors).collect.head
+    val numMachineCores = spark.range(0, 1)
```
Comment (mhamilton723) [inline src/lightgbm/src/main/scala/LightGBMUtils.scala]: Looks good! @imatiach-msft is it possible to remove the numMachineCores calculation and just use the confCores calc?

Comment (imatiach-msft) [reply src/lightgbm/src/main/scala/LightGBMUtils.scala]: updated based on our discussion (put the conf code inside a try/catch)

**Pull request merged**